src/aiida_quantumespresso/workflows/xspectra/crystal.py

```diff
@@ -0,0 +1,676 @@
+                        'any wavefunction data.'
+                    )
+
+
+    def setup(self):
+        """Set required context variables."""
+        custom_elements_list = self.inputs.get('elements_list', None)
+        if not custom_elements_list:
+            structure = self.inputs.structure
+            self.ctx.elements_list = [kind.symbol for kind in structure.kinds]
+        else:
+            self.ctx.elements_list = custom_elements_list.get_list()
+
+        if 'core_wfc_data' in self.inputs.keys():
+            self.ctx.core_wfc_data = self.inputs.core_wfc_data
+
+
+    def should_run_relax(self):
+        """If the 'relax' input namespace was specified, we relax the input structure."""
+        return 'relax' in self.inputs
+
+    def run_relax(self):
+        """Run the PwRelaxWorkChain to run a relax PwCalculation."""
+        inputs = AttributeDict(self.exposed_inputs(PwRelaxWorkChain, namespace='relax'))
+        inputs.metadata.call_link_label = 'relax'
+        inputs.structure = self.inputs.structure
+
+        running = self.submit(PwRelaxWorkChain, **inputs)
+
+        self.report(f'launching PwRelaxWorkChain<{running.pk}>')
+
+        return ToContext(relax_workchain=running)
+
+    def inspect_relax(self):
+        """Verify that the PwRelaxWorkChain finished successfully."""
+        workchain = self.ctx.relax_workchain
+
+        if not workchain.is_finished_ok:
+            self.report(f'PwRelaxWorkChain failed with exit status {workchain.exit_status}')
+            return self.exit_codes.ERROR_SUB_PROCESS_FAILED_RELAX
+
+        relaxed_structure = workchain.outputs.output_structure
+        self.ctx.relaxed_structure = relaxed_structure
+        self.out('optimized_structure', relaxed_structure)
+
+    def get_xspectra_structures(self):
+        """Perform symmetry analysis of the relaxed structure and get all marked structures for XSpectra."""
+        from aiida_quantumespresso.workflows.functions.get_xspectra_structures import get_xspectra_structures
+
+        elements_list = orm.List(self.ctx.elements_list)
+
+        inputs = {
+            'absorbing_elements_list' : elements_list,
+            'absorbing_atom_marker' : self.inputs.abs_atom_marker,
+            'metadata' : {
+                'call_link_label' : 'get_xspectra_structures'
+            }
+        }
+        if 'structure_preparation_settings' in self.inputs:
+            optional_cell_prep = self.inputs.structure_preparation_settings
+            for key, value in optional_cell_prep.items():
+                if isinstance(value, float):
+                    node_value = orm.Float(value)
+                if isinstance(value, int):
+                    node_value = orm.Int(value)
+                if isinstance(value, bool):
+                    node_value = orm.Bool(value)
+                if isinstance(value, str):
+                    node_value = orm.Str(value)
```

Comment:
Creating new node data here causes a loss of data provenance.

Reply:
Bear in mind that `structure_preparation_settings`, which we are unpacking here, is already an input node to the WorkChain. We're only losing provenance in the technical sense that the node identity is being lost, not the actual information itself.

I will see if I can refactor `structure_preparation_settings` into an `input_namespace`, which would then preserve the IDs of these parameters, but I'm not sure how well this works for different data types.